### ISSUE TITLE
golang: standardise "version" output

### ISSUE THREAD

**Issue by jonboulle** [coreos/docs#637]
As discussed [here](https://github.com/coreos/flannel/pull/349/files#r42780761), we should come up with a standard way for our Go executables to express their versions. I suggest erring on the side of verbosity and starting from something like etcd's current:

```
etcd Version: 2.2.1+git
Git SHA: Not provided (use ./build instead of go build)
Go Version: go1.5.1
Go OS/Arch: linux/amd64
```

**Comment by crawford** [comment]
I would prefer using `git describe` to derive the version number of applications (Go libraries might be another story given `go get`). In etcd's case, the version is hard-coded and the Git SHA is derived. I think we can drop the OS/Arch since it is redundant (if the binary isn't built for your system, it isn't going to run). Maybe this is sufficient:

```
foo Version: 1.0.1
Go Version: go1.5.1
```

**Comment by jonboulle** [comment]
@crawford what do you suggest for the `go get` case? I like that etcd's is a little more specific than our `0.0.0+you-done-did-build-wrong`

**Comment by crawford** [comment]
Well, if it is a library, I think you have to include the version number since `go get` doesn't pull the git history. What do you mean by "a little more specific"? The reason fleet has any version number at all is because it won't build without it (which I think is dumb). The other projects use "was not built correctly" as the version string.

**Comment by jonboulle** [comment]
> The reason fleet has any version number at all is because it won't build without it

I don't know what you mean by this, please explain

> What do you mean by "a little more specific"? 

The idea is that if the version is hard-coded to `2.2.0+git`, but then overridden by the build script with `git describe`, then you get the best of both worlds - a non-specific but correct version for `go get`, and an exact version if the build script is used.

**Comment by crawford** [comment]
>> The reason fleet has any version number at all is because it won't build without it

> I don't know what you mean by this, please explain

Because of this [panic](https://github.com/coreos/fleet/blob/2d024517a2c9f1e7492ef00f9ac26500b27556b2/version/version.go#L32) in an `init()`, the version number must always be valid when compiling. Because TravisCI doesn't exposed the git environment, the tests failed without that hack. The other projects don't suffer from this because they don't require a valid version number.

**Comment by jonboulle** [comment]
I don't understand how that's relevant/concerning. ignition still has a hard-coded version string.. unless you're really quibbling about "was not built properly" vs "0.0.0+was-not-build-properly"...

**Comment by crawford** [comment]
Well, it is relevant because it doesn't make any sense to bother with `git describe` if you are using a hard-coded string. Otherwise, you have info duplicated between your revision control and code. I never update the version string in Ignition because it will never change. Your proposal makes it such that the version number _will_ need to be updated.

**Comment by jonboulle** [comment]
> Well, it is relevant because it doesn't make any sense to bother with git describe if you are using a hard-coded string. Otherwise, you have info duplicated between your revision control and code.

Assuming that the most precise version information possible is always desirable, is my example really not sufficient to convince you of this?  It's an extremely modest amount of duplication and trivially taken care of by a scripted release process.

**Comment by crawford** [comment]
I feel like we are talking past one another...

What is the point of using git describe? You already have the version number hard-coded from your release process.

**Comment by jonboulle** [comment]
How about we take this to the ping pong table.

**Comment by crawford** [comment]
As long as I can bring my shake.

**Comment by crawford** [comment]
@jonboulle alerted me to the pagan belief in `go get`ing an application. As long as people are doing this, there _must_ be a version number in code. Rewriting the version number with the output of `git describe` then gives you added specificity when building directly. LGTM.

**Comment by jonboulle** [comment]
tl;dr: Either we should intentionall b

**Comment by eyakubovich** [comment]
I don't particularly care for `go get` but it must work if a tagged release was fetched from GH (or elsewhere) in a form of a src tarball. And that means that we need to include version in the source code.

**Comment by eyakubovich** [comment]
@crawford Until this realization, I was with you that we should only rely on `git describe`

**Comment by jonboulle** [comment]
Perhaps of interest: https://github.com/gravitational/version

**Comment by ElijahCaine** [comment]
Closing due to inactivity.
